File: apps/admin/src/models/link.ts
import type { PaginateResult } from "./base";

export enum LinkType {
  Friend,
  Collection,
}

export enum LinkState {
  Pass,
  Audit,
  Outdate,
  Banned,
  Rejected,
}

export interface LinkModel {
  _id: string;
  created: string;
  name: string;
  url: string;
  avatar: string;
  description?: string;
  type: LinkType;
  state: LinkState;
  email?: string | null;
  rssurl?: string | null;
  techstack?: string[] | null;
  health?: LinkHealthStatus | null;
}

export type LinkResponse = PaginateResult<LinkModel>;

export interface LinkHealthStatus {
  link_id: string;
  url: string;
  is_alive: boolean;
  status_code?: number;
  latency_ms?: number;
  hosting_provider: string;
  checked_at: string;
  error_message?: string;
  is_stale: boolean;
}

export interface LinkStateCount {
  audit: number;
  pass: number;
  outdate: number;
  banned: number;
  rejected: number;
}

export const LinkStateNameMap: Record<keyof typeof LinkState, string> = {
  Audit: "待审核",
  Pass: "通过",
  Outdate: "过时",
  Banned: "屏蔽",
  Rejected: "不通过",
};

File: apps/backend/src/models/link.rs
//! Link (Friend) model

use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

use super::serializers::{deserialize_id_to_string, serialize_datetime};

/// Link state constants
/// - 0: Normal
/// - 1: Pending
pub struct LinkState;

impl LinkState {
    pub const NORMAL: i32 = 0;
    pub const PENDING: i32 = 1;
}

/// Link type constants
/// - 0: Friend (default)
pub struct LinkType;

impl LinkType {
    pub const FRIEND: i32 = 0;
}

/// Link (Friend) model
#[derive(Debug, Serialize, Deserialize, Clone, ToSchema)]
pub struct Link {
    #[serde(rename = "_id", deserialize_with = "deserialize_id_to_string")]
    #[schema(value_type = String)]
    pub id: String,
    pub name: String,
    pub url: String,
    #[serde(default)]
    pub avatar: String,
    #[serde(default)]
    pub description: String,
    #[serde(default)]
    pub state: i32,
    #[serde(default)]
    pub r#type: i32,
    #[serde(
        serialize_with = "serialize_datetime",
        deserialize_with = "crate::models::serializers::deserialize_flexible_datetime"
    )]
    #[schema(value_type = String)]
    pub created: bson::DateTime,
    pub email: Option<String>,
    pub rssurl: Option<String>,
    pub techstack: Option<Vec<String>>,
}

/// Link apply request
#[derive(Debug, Deserialize, ToSchema)]
pub struct LinkApplyRequest {
    pub name: String,
    pub url: String,
    pub avatar: String,
    pub description: String,
    pub email: String,
    pub code: String,
    pub rssurl: Option<String>,
    pub techstack: Option<Vec<String>>,
}

/// Send verification code request
#[derive(Debug, Deserialize, ToSchema)]
pub struct SendCodeRequest {
    pub email: String,
}

/// Link with health status (for API responses)
#[derive(Debug, Serialize, Deserialize, Clone, ToSchema)]
pub struct LinkWithHealth {
    #[serde(flatten)]
    pub link: Link,
    pub health: Option<LinkHealthStatus>,
}

/// Link health status
#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct LinkHealthStatus {
    /// Link ID
    pub link_id: String,
    /// Link URL
    pub url: String,
    /// Is alive
    pub is_alive: bool,
    /// HTTP status code
    pub status_code: Option<u16>,
    /// Latency in milliseconds
    pub latency_ms: Option<u64>,
    /// Hosting provider (lowercase string, e.g. "cloudflare", "vercel")
    #[serde(default)]
    pub hosting_provider: String,
    /// Checked at
    pub checked_at: String,
    /// Error message
    pub error_message: Option<String>,
    /// Is stale data (being refreshed in background)
    pub is_stale: bool,
}
